examples/riot/coap_channel/coap_channel.c

```diff
@@ -0,0 +1,284 @@
+  // coap_opt_add_format(pdu, COAP_FORMAT_TEXT);
+  // size_t resp_len = coap_opt_finish(pdu, COAP_OPT_FINISH_PAYLOAD);
+
+  // if (self->connected_to_client) {
+  //   // This server is already connected to a client
+  //   return false;
+  // }
+
+  // // Set as connected
+  // self->connected_to_client = true;
+
+  // // Connect successful
+  // return true;
+  gcoap_resp_init(pdu, buf, len, COAP_CODE_CONTENT);
+  coap_opt_add_format(pdu, COAP_FORMAT_TEXT);
+  size_t resp_len = coap_opt_finish(pdu, COAP_OPT_FINISH_PAYLOAD);
+
+  /* write the RIOT board name in the response buffer */
+  if (pdu->payload_len >= strlen(RIOT_BOARD)) {
+    memcpy(pdu->payload, RIOT_BOARD, strlen(RIOT_BOARD));
+    return resp_len + strlen(RIOT_BOARD);
+  } else {
+    puts("gcoap_cli: msg buffer too small");
+    return gcoap_response(pdu, buf, len, COAP_CODE_INTERNAL_SERVER_ERROR);
+  }
+}
+
+static ssize_t _server_disconnect_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, coap_request_ctx_t *ctx) {
+  LF_DEBUG(NET, "CoapChannel: Server disconnect handler");
+  CoapChannel *self = _get_coap_channel_by_remote(ctx->remote);
+
+  gcoap_resp_init(pdu, buf, len, COAP_CODE_CONTENT);
+  coap_opt_add_format(pdu, COAP_FORMAT_TEXT);
+  size_t resp_len = coap_opt_finish(pdu, COAP_OPT_FINISH_PAYLOAD);
+
+  // Set as disconnect
+  self->state = NETWORK_CHANNEL_STATE_DISCONNECTED;
+
+  // Disconnect successful
+  return true;
+}
+
+static ssize_t _server_message_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, coap_request_ctx_t *ctx) {
+  LF_DEBUG(NET, "CoapChannel: Server message handler");
+  CoapChannel *self = _get_coap_channel_by_remote(ctx->remote);
+
+  gcoap_resp_init(pdu, buf, len, COAP_CODE_CONTENT);
+  coap_opt_add_format(pdu, COAP_FORMAT_TEXT);
+  size_t resp_len = coap_opt_finish(pdu, COAP_OPT_FINISH_PAYLOAD);
+
+  // TODO
+
+  return true;
+}
+
+static const coap_resource_t _resources[] = {
+    {"/connect", COAP_GET, _server_connect_handler, NULL},
+    {"/disconnect", COAP_GET, _server_disconnect_handler, NULL},
+    {"/message", COAP_GET, _server_message_handler, NULL},
+};
+
+static gcoap_listener_t _listener = {&_resources[0], ARRAY_SIZE(_resources), GCOAP_SOCKET_TYPE_UDP, NULL, NULL, NULL};
+
+static lf_ret_t CoapChannel_open_connection(NetworkChannel *untyped_self) {
+  LF_DEBUG(NET, "CoapChannel: Open connection");
+  CoapChannel *self = (CoapChannel *)untyped_self;
+
+  /* Server */
+  // Do nothing
+
+  /* Client */
+  // Do nothing
+
+  self->state = NETWORK_CHANNEL_STATE_OPEN;
+}
+
+static void _client_try_connect_callback(const gcoap_request_memo_t *memo, coap_pkt_t *pdu,
+                                         const sock_udp_ep_t *remote) {
+  LF_DEBUG(NET, "CoapChannel: Try connect callback");
+  CoapChannel *self = _get_coap_channel_by_remote(remote);
+
+  if (memo->state == GCOAP_MEMO_TIMEOUT || coap_get_code_class(pdu) != COAP_CLASS_SUCCESS) {
+    self->state = NETWORK_CHANNEL_STATE_DISCONNECTED;
```

Comment:
Do we need to acquire a lock before modifying this variable? THis might be called from a different thread?

Reply:
I need to investigate this I am not sure about the threading situation in gcoap right now